be/src/olap/rowset/segment_v2/column_zone_map.h

```diff
@@ -0,0 +1,166 @@
+            _null_flag(false), _non_null_flag(false) {
+        PageBuilderOptions options;
+        options.data_page_size = 0;
+        _page_builder.reset(new BinaryPlainPageBuilder(options));
+        _field.reset(Field::create_by_type(_type_info->type()));
+        _max_string_value = _arena.Allocate(OLAP_STRING_MAX_LENGTH);
+        _min_value = _arena.Allocate(_type_info->size());
+        // we should allocate max varchar length and set to max for min value
+        Slice* min_slice = (Slice*)_min_value;
+        min_slice->data = _max_string_value;
+        min_slice->size = OLAP_STRING_MAX_LENGTH;
+        _field->set_to_max(_min_value);
+        _max_value = _arena.Allocate(_type_info->size());
+        _field->set_to_min(_max_value);
+    }
+
+    Status add(const uint8_t* vals, size_t count) {
+        if (vals != nullptr) {
+            for (int i = 0; i < count; ++i) {
+                if (_field->compare(_min_value, (char*)vals) > 0) {
+                    _field->deep_copy_content(_min_value, (const char*)vals, &_arena);
+                }
+                if (_field->compare(_max_value, (char*)vals) < 0) {
+                    _field->deep_copy_content(_max_value, (const char*)vals, &_arena);
+                }
+                vals += _type_info->size();
+                if (!_non_null_flag) {
+                    _non_null_flag = true;
+                }
+            }
+        } else {
+            if (!_null_flag) {
+                _null_flag = true;
+            }
+        }
+        return Status::OK();
+    }
+
+    Status flush() {
+        ZoneMapPB page_zone_map;
+        page_zone_map.set_min(_field->to_string(_min_value));
+        page_zone_map.set_max(_field->to_string(_max_value));
+        page_zone_map.set_null_flag(_null_flag);
+        page_zone_map.set_non_null_flag(_non_null_flag);
+        std::string serialized_zone_map;
+        bool ret = page_zone_map.SerializeToString(&serialized_zone_map);
+        if (!ret) {
+            return Status::InternalError("serialize zone map failed");
+        }
+        Slice data(serialized_zone_map.data(), serialized_zone_map.size());
+        size_t num = 1;
+        RETURN_IF_ERROR(_page_builder->add((const uint8_t*)&data, &num));
+        // reset the variables
+        // we should allocate max varchar length and set to max for min value
+        Slice* min_slice = (Slice*)_min_value;
+        min_slice->data = _max_string_value;
+        min_slice->size = OLAP_STRING_MAX_LENGTH;
+        _field->set_to_max(_min_value);
+        _field->set_to_min(_max_value);
+        _null_flag = false;
+        _non_null_flag = false;
+        return Status::OK();
+    }
+
+    Slice finish() {
+        return _page_builder->finish();
+    }
+
+private:
+    const TypeInfo* _type_info;
+    std::unique_ptr<BinaryPlainPageBuilder> _page_builder;
+    std::unique_ptr<Field> _field;
+    // memory will be managed by arena
+    char* _min_value;
```

Comment:
I think you should create a struct to contain min_value max_value and flags. which may be used to condition compute

Reply:
I will add struct ZoneMap